### ISSUE TITLE
16bit per channel PNG images are incorrectly imported by the content-pipeline tool

### ISSUE THREAD

**Issue by roy-t** [MonoGame/MonoGame#6424]
I believe PNG images that have 16 bits per channel are incorrectly processed by the content-pipeline tool. 

I've created an extremely minimal reproduction here: https://github.com/roy-t/monogame_16bitpp_png_bug

Below you see two textures rendered with the spritebatch, with default settings. The left sprite is a texture generated from an 8 bit per pixel PNG file, and is displayed correctly. The right sprite is a texture generated from the same PNG file, this time stored with 16 bits per pixel. As you can see the colors are completely different.

![bug](https://user-images.githubusercontent.com/623634/44863637-f371a180-ac7d-11e8-8532-a2b4835fd206.png)

#### What version of MonoGame does the bug occur on:
- MonoGame 3.6.0.1625

#### What operating system are you using:
- Windows

#### What MonoGame platform are you using:
- Windows, DirectX

**Comment by roy-t** [comment]
I think the bug is in the code below.

Although the code does check the bytes per pixel to see how far to jump to get to the next color channel. It assumes a single color is one byte. If I'd patch this code to down-sample from however many bits to a single byte, would that be accept as a PR?

https://github.com/MonoGame/MonoGame/blob/6f34eb393aa0ac005888d74c5c4c6ab5615fdc8c/MonoGame.Framework/Utilities/Png/PngReader.cs#L245
```
{
    switch (colorType)
    {
        // snip
        case ColorType.Rgb:
            for (int x = 0; x < width; x++)
            {
                int offset = 1 + (x * bytesPerPixel);
                
                int red = defilteredScanline[offset];
                int green = defilteredScanline[offset + bytesPerSample];
                int blue = defilteredScanline[offset + 2 * bytesPerSample];

                data[(y * width) + x] = new Color(red, green, blue);
            }

            break;
    // snip
    }
}

```

**Comment by tomspilman** [comment]
Thanks for the detailed report @roy-t !

I don't know this code as well... we would totally accept a PR to fix it along with a unit test to verify the fix.

**Comment by roy-t** [comment]
I first found the PngReader by searching for PNG, but not it seems that, at least on Windows, its not referenced. Using the unit tests I came to this file:

https://github.com/MonoGame/MonoGame/blob/6f34eb393aa0ac005888d74c5c4c6ab5615fdc8c/MonoGame.Framework.Content.Pipeline/TextureImporter.cs#L139

Which is very suspect, why would the R and B channels of the FREE_IMAGE_TYPE.FIT_RGB16 type be switched but those of the FREE_IMAGE_TYPE.FIT_RGBA16 not? I quickly checked in paint.net and indeed if you switch the R and B channels the images look the same.

I tried to make the change, but I must be doing something wrong the content pipeline tool, because I keep getting the same output. :/. I'll investigate further in the weekend.

**Comment by roy-t** [comment]
@tomspilman  I've figured it out. In the screenshot you now see 8 bit RGB, 8 bit RGBA, 16 bit RGB and 16 bit RGBA png files.

The fix was that the red and blue color channels should not be swapped for all 16 bit RGB(A) images. While it was not swapped for the 16 bit RGB images, and was swapped for the 16 bit RGBA. 

The same bug seems to exist for the FIT_RGBF/FIT_RGBAF (32 bit per pixel) image type. Do you happen to know what program I could use to generate 32 bit per channel images, so I can also test this case?

If all goes well I'll make a proper PR with this fix on Sunday. 

A happy picture :).
![capture](https://user-images.githubusercontent.com/623634/44895319-1a6aba80-acf4-11e8-8cee-5ad1012f4361.PNG)

**Comment by roy-t** [comment]
I've added the pull request. 💃

**Comment by Jjagg** [comment]
Fixed in #6425!
